certora/specs/HashConsensus.spec

```diff
@@ -0,0 +1,793 @@
+// 21. submitReport() - revert if same oracle reports the same slot+hash (cannot double vote for same report)
+// 22. submitReport() - single call to submit report increases only one support for one variant
+//                      sum of supports of all variants < total members,
+//                      because if Oracle member changes its mind, its previous support is removed
+// 23. submitReport() - all variants should be removed when new frame starts
+
+
+//  1. All the external view functions cannot revert under any circumstance,
+//     especially getCurrentFrame() as it is used by LegacyOracle and BaseOracle
+// Status: Fail
+// https://vaas-stg.certora.com/output/80942/985192ce8baf4725a8b6d29c1d0dc2af/?anonymousKey=f2574e5803ece9f5b8d734ecd68b1d9ac714c1d5
+rule viewFunctionsDoNotRevert(method f)
+    filtered { f -> f.isView }
+{
+    env e; calldataarg args;
+
+    require e.msg.value == 0; // view functions revert is you send eth
+
+    f@withrevert(e,args);
+    assert !lastReverted;
+}
+
+/// POTENTIAL ISSUE: SLOTS_PER_EPOCH, SECONDS_PER_SLOT, GENESIS_TIME can all be set to zero at constructor
+/// if the above are zero => getCurrentFrame() will revert
+// focus only on getCurrentFrame() and verify it does not revert
+// Status: Timeout! (without simplification)
+// https://vaas-stg.certora.com/output/80942/899cc6d765c7467c9e2af018ccca2ca5/?anonymousKey=ca1152175e66f3e61e7d2bffaded5098c62c3bbe
+// Status: Pass (with simplification)
+// https://vaas-stg.certora.com/output/80942/01ca967565ee4fc58101e7037160aaa1/?anonymousKey=fd302ac936356aa223eb706d466eeae5de7a155f
+rule getCurrentFrameDoesNotRevert() {
+    env e; calldataarg args;
+
+    require e.msg.value == 0;                       // view functions revert is you send eth
+    require e.block.timestamp > 1672531200;         // 01.01.2023 00:00:00
+    require e.block.timestamp < 2524608000;         // 01.01.2050 00:00:00
+
+    uint256 slotsPerEpoch; uint256 secondsPerSlot; uint256 genesisTime;
+    slotsPerEpoch, secondsPerSlot, genesisTime = getChainConfig(e);
+
+    require slotsPerEpoch > 0;                      // must be required at constructor
+    require slotsPerEpoch == 1;                     // simplification
+    require secondsPerSlot > 0;                     // must be required at constructor
+    require secondsPerSlot == 1;                    // simplification
+    require genesisTime < e.block.timestamp;        // must be required at constructor
+
+    uint256 initialEpoch; uint256 epochsPerFrame; uint256 fastLaneLengthSlots;
+    initialEpoch, epochsPerFrame, fastLaneLengthSlots = getFrameConfig(e);
+    require epochsPerFrame > 0;                     // constructor already ensures this
+    require epochsPerFrame < 31536000;              // assuming less than 1 year per frame
+
+    // assuming correct configuration of the frame, otherwise revert
+    require initialEpoch < (e.block.timestamp - genesisTime) / (secondsPerSlot * slotsPerEpoch);
```

Comment:
Can it be proved as an invariant, given that every call is at timestamp > genesis?